Adding support of KV Patch operation

yfaney opened the pull request:
# Issue Link
[Support for KV Patch in VaultTemplate #585](https://github.com/spring-projects/spring-vault/issues/585)

# Motivation and Context 
This change will enable the support of KV patch operation in VaultKeyValue2Template.
Note that this command will only work with Vault Secret Engine V2.
During the change, I tweaked a doRead() method to get the metadata properly in V2 access which is required to perform patch operation.
Please feel free to suggest improvement of this code. I tried the minimum change so it may have a room to refactor.
This change is based on the source code of the vault CLI:
https://github.com/hashicorp/vault/blob/master/command/kv_patch.go#L66

# How to test this change
Performing mvn install

# Test Evidence
Please see the attached screenshot
![Screen Shot 2020-09-24 at 12 52 04 AM](https://user-images.githubusercontent.com/5394800/94117449-104b1580-fe01-11ea-84f8-ebe662e2988f.png)

pivotal-issuemaster commented:
@yfaney Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-vault&pullRequestId=587)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-vault&pullRequestId=587) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@yfaney Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-vault&pullRequestId=587)!

mp911de commented on spring-vault-core/src/main/java/org/springframework/vault/core/VaultKeyValue2Template.java:
```diff
@@ -89,4 +90,29 @@ public void put(String path, Object body) {
+	 * @param kv New key value map to be updated
+	 * @since 2.3
+	 */
+	public void patch(String path, Map<String, ?> kv) {
```
This operation should be also provided for the version 1 backend. Since this class is package-protected, there's no way to consume this method. Therefore, the method should be also made available in the `VaultKeyValueOperations` interface.

yfaney replied:
Applied:
1. VaultKeyValueOperations.java : https://github.com/spring-projects/spring-vault/pull/587/commits/95d25bd9120c340cecc2eab71c67ffc047e41f07#diff-93f96dc2bd921937c3f5ee0eda355c0dR70
1. VaultKeyValue2Template.java : https://github.com/spring-projects/spring-vault/pull/587/commits/95d25bd9120c340cecc2eab71c67ffc047e41f07#diff-917f198074796659086e92ad34820ddfR101
1. VaultKeyValue1Template.java : https://github.com/spring-projects/spring-vault/pull/587/commits/95d25bd9120c340cecc2eab71c67ffc047e41f07#diff-36bdc62ca9e4f9e53d696f8675e326eaR111

mp911de commented on spring-vault-core/src/main/java/org/springframework/vault/core/VaultKeyValue2Template.java:
```diff
@@ -89,4 +90,29 @@ public void put(String path, Object body) {
+
+		// To do patch operation, we need to do a read operation first
+		VaultResponse readResponse = get(path);
+		Assert.notNull(readResponse, "VaultResponse must not be null");
```
If the secret doesn't exist, a meaningful exception should be thrown. I think a new `SecretNotFoundException` type (`extends VaultException`) would make sense.

yfaney replied:
Applied here: https://github.com/spring-projects/spring-vault/pull/587/commits/95d25bd9120c340cecc2eab71c67ffc047e41f07#diff-917f198074796659086e92ad34820ddfR106-R112

mp911de commented on spring-vault-core/src/test/java/org/springframework/vault/core/VaultKeyValueTemplateVersionedIntegrationTests.java:
```diff
@@ -35,4 +45,28 @@
+
+		Map<String, String> newSecret = Collections.singletonMap(newKey, "newValue");
+
+		((VaultKeyValue2Template) this.kvOperations).patch(key, newSecret);
```
Once the `patch` method gets exposed through `VaultKeyValueOperations`, there's no need to have the downcast.

yfaney replied:
Now it does not downcast: https://github.com/spring-projects/spring-vault/pull/587/commits/95d25bd9120c340cecc2eab71c67ffc047e41f07#diff-98a440907d66c701b5a12de4a5db680aR61

mp911de commented on spring-vault-core/src/main/java/org/springframework/vault/core/VaultKeyValue2Template.java:
```diff
@@ -89,4 +90,29 @@ public void put(String path, Object body) {
+		body.put("data", data);
+		body.put("options", Collections.singletonMap("cas", metadata.get("version")));
+
+		doWrite(createDataPath(path), body);
```
There should be also a reaction on whether the write was successful. A concurrent process could have updated the secret in the meantime. Returning `true` on success (similar to Java's `compareAndSet` methods in the concurrency util package) would make sense.

yfaney replied:
I changed the return type as boolean and true if the response is successful, but please let me know if you want another return type such as VaultResponse itself or something else:
https://github.com/spring-projects/spring-vault/pull/587/commits/95d25bd9120c340cecc2eab71c67ffc047e41f07#diff-917f198074796659086e92ad34820ddfR120-R125

And FYI, according to the document of HashCorp's KV secret engine V2, we will be safe from the concurrency issue because Vault will not allow if the secret's versioning is not correct.(If the existing version of the secret is 3 and incoming one is still 3 it won't allow writing operation so it can avoid from the race condition)
Please see the following links:
https://learn.hashicorp.com/tutorials/vault/versioned-kv
https://www.vaultproject.io/docs/commands/kv/patch

mp911de reviewed (changes requested):
Thanks a lot. I had a look and left a few comments.

yfaney commented:
Hi, @mp911de I addressed all of your comments above. Please review my changes again and let me know if there are still something to be changed. Thanks!

mp911de reviewed (approved):
The formatting is a bit off and license headers are missing. That's not a blocker, just wanted to mention it and we will take care of this during the merge.

Going to take this PR from here.

yfaney commented:
@mp911de awesome, I'll keep that in mind and try following the formatting and license headers in the next time :)
